pkg/agent/client/nodeconn.go

```diff
@@ -0,0 +1,64 @@
+package client
+
+import (
+	"sync"
+
+	"google.golang.org/grpc"
+)
+
+type nodeConn struct {
+	conn       *grpc.ClientConn
+	refcount   int32
+	mu         sync.RWMutex
+	alive      bool
+	isReleased bool
+}
+
+func newNodeConn(conn *grpc.ClientConn) *nodeConn {
+	return &nodeConn{
+		alive:    true,
+		conn:     conn,
+		refcount: 1,
+	}
+}
+
+func (c *nodeConn) AddRef() {
+	c.mu.Lock()
+	c.refcount++
+	c.mu.Unlock()
+}
+
+func (c *nodeConn) Release() {
+	c.mu.Lock()
+	c.release()
+	c.mu.Unlock()
+}
+
+func (c *nodeConn) release() {
+	c.refcount--
+	if c.refcount == 0 {
+		c.conn.Close()
+		c.isReleased = true
```

Comment:
nit: maybe c.conn can be set to nil to convey that it has been "released" instead of using a separate variable to track it?

Reply:
+1